webrtc/peerconnection.go

```diff
@@ -0,0 +1,58 @@
+package webrtc
+
+import (
+	"github.com/pion/webrtc/v3"
+)
+
+// PeerConnection represents RTCPeerConnection.
+type PeerConnection interface {
+	// AddTransceiverFromTrack creates a new RTPTransceiver from LocalTrack
+	// and register it to the PeerConnection.
+	// RTPTransceiver represents a bidirectional stream.
+	//
+	// ref: https://developer.mozilla.org/en-US/docs/Web/API/RTCPeerConnection/addTransceiver
+	AddTransceiverFromTrack(LocalTrack, ...RTPTransceiverInit) (RTPTransceiver, error)
+	// AddTransceiverFromTrack creates a new RTPTransceiver from RTPCodecType
+	// and register it to the PeerConnection.
+	// RTPTransceiver represents a bidirectional stream.
+	//
+	// ref: https://developer.mozilla.org/en-US/docs/Web/API/RTCPeerConnection/addTransceiver
+	AddTransceiverFromKind(webrtc.RTPCodecType, ...RTPTransceiverInit) (RTPTransceiver, error)
+}
+
+// RTPTransceiverInitDict represents RTCRtpTransceiverInit dictionary.
+type RTPTransceiverInitDict struct {
+	Direction     webrtc.RTPTransceiverDirection
+	SendEncodings []RTPParameters
+}
+
+// RTPTransceiverInit is a functional option type to set RTPTransceiverInitDict to AddTransceiverFromKind/Track.
```

Comment:
I don't think we should do `RTPTransceiverInit`. I would go with a functional API if we weren't trying to match the browser. 

* We have two ways to do one thing. People will try to merge examples/other code and may be confused.
* It doesn't match the W3C API.
* In the future the browser API may change/expand and if we don't follow closely it might put us in a bad spot.

Reply:
I also don't think it's the best way, but `AddTransceiver(trackOrKind RTPCodecType, init ...RtpTransceiverInit) (*RTPTransceiver, error)` seems very problemistic to me.

- On v2 API, `AddTransceiver(trackOrKind RTPCodecType, init ...RtpTransceiverInit) (*RTPTransceiver, error)`, user can set `RtpTransceiverInit` multiple times and the may have multiple meanings (first one is valid?, last one is valid?, or all inits will be merged?)
- Passing Go struct can't handle default value properly
  - Can't recognize difference between `unspecified` and `empty array`
  - Pointer to the primitive types are painful

> - We have two ways to do one thing. People will try to merge examples/other code and may be confused.

User can only pass functional option to the function. Basically no other way to specify the option.

> - It doesn't match the W3C API.
> - In the future the browser API may change/expand and if we don't follow closely it might put us in a bad spot.

I think our implementation already differ from WebAPI. In WebAPI there are no `AddTransceiverFromKind/AddTransceiverFromTrack` but only `AddTransceiver`.
In WebAPI, `RTCRtpTransceiverInit` is `dict`, so the function can recognize `unspecified` and `empty array`.
If we strictly match W3C API, our API must be
```go
AddTransceiver(trackOrKind interface{}, init map[string]interface{}) (*RTPTransceiver, error)
```

Inside `AddTransceiverFromKind/AddTransceiverFromTrack`, array of `RTPTransceiverInit` are quickly translated into `RTPTransceiverInitDict` with default values. It's easy to translated into WebAPI and highly extensible than directly using struct.
```go
initDict := RTPTransceiverInitDict { // default values }
for _, i := range init {
  i(initDict)
}
```

Reply:
What about.

```
AddTransceiver(trackOrKind interface{}, init ...RtpTransceiverInit) (*RTPTransceiver, error)
```

* Maybe we can figure out a way to do a real interface instead of `interface{}`
* What do you think of doing `len(init)`? It will just be hard to mentally map `map[string]interface{}` back to Javascript API.

Reply:
@at-wat What do you think of ^ ?

I think this is the last needed API before I can start making examples up

Reply:
If you feel strongly about the functional route we can do that! I think we need to make the entire PeerConnection API functional then.

I think having two API patterns is worse then having one bad one.

Reply:
I don't have very strong opinion especially in this case.
(I think it's better to use functional options in SettingEngine to handle default setting properly.)
For `RTPTransceiverInit` case, making it
```
AddTransceiver*(..., init *RTPTransceiverInit) (*RTPTransceiver, error)
```
and require `nil` as a second argument maybe cleaner.

On the other hand, I feel receiving `interface{}` is not very clean and better to have strict typing if possible.

Reply:
Ok that sounds good. The following works for me.

```
	AddTransceiverFromTrack(LocalTrack, *RTPTransceiverInit) (RTPTransceiver, error)
	AddTransceiverFromKind(webrtc.RTPCodecType, *RTPTransceiverInit) (RTPTransceiver, error)
```

I am also really in support of making everything functional! If you want to start moving the entire code base over. The most important thing to me is that everyone feels happy and has ownership. I am convinced it is a good idea, just don't want to see the code base have multiple patterns.

It is way better to have someone in this stuff that is passionate :)

Reply:
Created an issue about config: https://github.com/pion/webrtc-v3-design/issues/16